redfish-core/lib/power_supply_metrics.hpp

```diff
@@ -0,0 +1,366 @@
+                            const std::string& averagePath,
+                            const std::string& maximumPath)
+{
+    BMCWEB_LOG_DEBUG << "Get Values from serviceName: " << serviceName
+                     << " objectPath: " << averagePath
+                     << " interfaceName: " << averageInterface;
+
+    crow::connections::systemBus->async_method_call(
+        [aResp, serviceName,
+         maximumPath](const boost::system::error_code ec2,
+                      const std::variant<averageMaxArray>& intfValues) mutable {
+            if (ec2)
+            {
+                BMCWEB_LOG_DEBUG << "D-Bus response error";
+                messages::internalError(aResp->res);
+                return;
+            }
+
+            const averageMaxArray* intfValuesPtr =
+                std::get_if<averageMaxArray>(&intfValues);
+            if (intfValuesPtr == nullptr)
+            {
+                messages::internalError(aResp->res);
+                return;
+            }
+
+            // Populate below with data read from D-Bus.
+            averageMaxArray averageValues;
+            for (const auto& values : *intfValuesPtr)
+            {
+                // The first value returned is the timestamp, it is in
+                // milliseconds since the Epoch.
+                auto dateTime = std::get<0>(values);
+                // The second value returned is in watts. It is the average
+                // watts this power supply has used in a 30 second interval.
+                auto value = std::get<1>(values);
+
+                BMCWEB_LOG_DEBUG
+                    << "DateTime: "
+                    << crow::utility::getDateTime(
+                           static_cast<std::time_t>(dateTime / 1000));
+                BMCWEB_LOG_DEBUG << "Values: " << value;
+
+                averageValues.emplace_back(dateTime, value);
+            }
+
+            getMaximumValues(aResp, serviceName, maximumPath, averageValues);
+        },
+        serviceName, averagePath, "org.freedesktop.DBus.Properties", "Get",
+        averageInterface, "Values");
+}
+
+/**
+ * @brief Get power supply average, maximum and date values given chassis and
+ * power supply IDs.
+ *
+ * @param[in] aResp - Shared pointer for asynchronous calls.
+ * @param[in] chassisID - Chassis to which the values are associated.
+ * @param[in] powerSupplyID - Power supply to which the values are associated.
+ *
+ * @return None.
+ */
+inline void getValues(const std::shared_ptr<bmcweb::AsyncResp>& aResp,
+                      const std::string& chassisID,
+                      const std::string& powerSupplyID)
+{
+    BMCWEB_LOG_DEBUG
+        << "Get power supply date/average/maximum input power values";
+    // Setup InputPowerHistoryItem values array.
+    // It will have 0 to many date/timestamp, average, and maximum entries.
+    aResp->res.jsonValue["Oem"]["IBM"]["InputPowerHistoryItem"]["@odata.type"] =
```

Comment:
Oh, the validator must check against the XML not the JSON It is InputPowerHistoryItems in the JSON

Reply:
Recursive grep ... changing all InputPowerHistoryItem to InputPowerHistoryItems.